### PR TITLE
Dev/ncar candidate 2018 09 26

### PR DESCRIPTION
**Pull request by gustavo-marques**
This PR adds many capabilities, as summarized below. All answers in MOM6-examples are b4b with respect to the latest dev/master (only Intel was checked).
 
* Modularize MCT cap to facilitate comparison with the NOUPC cap that is currently under development; 
 
* Although there are many NOUPC-related commits here, the MOM6 NOUPC cap has been temporarily removed. It will be re-introduced on the next PR from dev/ncar to dev/master;

* Add safety checks for MOM_bkgnd_mixing schemes, including checking whether KD, KV and PRANDTL_BKGND are compatible;

* Fix call to shelf_calc_flux

* Add option to apply vertically uniform, latitude-dependent background diffusivity, as described in 
 Danabasoglu et al., 2012

* Add option to compute instantaneous melt potential;

* Make Hml from either KPP or ePBL available to the caps via surface state.

### CODE REVIEW

File: src/parameterizations/vertical/MOM_bkgnd_mixing.F90 [hunk 1]
```diff
@@ -196,21 +224,56 @@ subroutine bkgnd_mixing_init(Time, G, GV, param_file, diag, CS)
...
                                 CS%Henyey_IGW_background_new, &
                  "If true, use a better latitude-dependent scaling for the\n"//&
                  "background diffusivity, as described in \n"//&
                  "Harrison & Hallberg, JPO 2008.", default=.false.)
+  if (CS%Henyey_IGW_background_new) call check_bkgnd_scheme(CS, "HENYEY_IGW_BACKGROUND_NEW")
 
-  if (CS%Henyey_IGW_background .and. CS%Henyey_IGW_background_new) &
-     call MOM_error(FATAL, "set_diffusivity_init: HENYEY_IGW_BACKGROUND and \n"//&
-          "HENYEY_IGW_BACKGROUND_NEW are mutually exclusive. Set only one or none.")
+  if (CS%Kd>1.e-14 .and. (trim(CS%bkgnd_scheme_str)=="BRYAN_LEWIS_DIFFUSIVITY" .or.&
```
Comment (adcroft) [inline src/parameterizations/vertical/MOM_bkgnd_mixing.F90]: This constant should be removed. It's has dimensions and fails the dimensional re-scaling test. Moreover, the test should be >0, not >small number.

File: src/parameterizations/vertical/MOM_bkgnd_mixing.F90 [hunk 2]
```diff
@@ -159,10 +179,17 @@ subroutine bkgnd_mixing_init(Time, G, GV, param_file, diag, CS)
 
   call get_param(param_file, mdl, 'DEBUG', CS%debug, default=.False., do_not_log=.True.)
 
+  prandtl_bkgnd_default = Kv/CS%Kd
```
Comment (adcroft) [inline src/parameterizations/vertical/MOM_bkgnd_mixing.F90]: I think this line is why this PR is failing tests:
- This is yielding NaN or Infinity depending on the compiler and configuration.
- PRANDTL_BKGND is also being logged in the output in a format that cannot be read back in. 

I suspect the resolution is to only read/log PRANDTL_BKGND in those modes where you are actually using it (when it is not infinite).

Perhaps we should discuss this parameter in the dev call?

**Comment by kshedstrom** [comment]
As forecast, no OBC issues in this.

Kate

File: src/parameterizations/vertical/MOM_bkgnd_mixing.F90 [hunk 3]
```diff
@@ -159,10 +179,17 @@ subroutine bkgnd_mixing_init(Time, G, GV, param_file, diag, CS)
...
   call get_param(param_file, mdl, "PRANDTL_BKGND", CS%prandtl_bkgnd, &
                  "Turbulent Prandtl number used to convert vertical \n"//&
                  "background diffusivities into viscosities.", &
-                 units="nondim", default=1.0)
+                 units="nondim", default=prandtl_bkgnd_default)
+
+  if ( abs(Kv-CS%Kd*CS%prandtl_bkgnd)>1.e-14) then
```
[the remark this answers is not in the document]
Comment (alperaltuntas) [reply src/parameterizations/vertical/MOM_bkgnd_mixing.F90]: It is simply to account for floating point rounding error, e.g., ```0.1*3 /= 0.3```. I haven't thought about the dimensional re-scaling test though, so I will apply your following suggestion:

> the resolution is to only read/log PRANDTL_BKGND in those modes where you are actually using it (when it is not infinite).

**Comment by adcroft** [comment]
https://gitlab.gfdl.noaa.gov/ogrp/MOM6/pipelines/5951

**Comment by adcroft** [comment]
Any feedback on updates from @awallcraft, @jiandewang and @kshedstrom ?

**Comment by awallcraft** [comment]
I'm not setup to test pull requests.  This is OK with me, although I am interested in the MOM6 NUOPC cap.

**Comment by kshedstrom** [comment]
Alan, if the pull request is on one of our codes, we can go to github to
get the steps for downloading it. Since this isn't code we own, I find it
easiest to set up a git remote for each contributor I'm asked to evaluate.
Then I can just "git fetch gustavo" to bring it into my repo, gustavo being:

gustavo https://github.com/gustavo-marques/MOM6


The latest came in on this branch: gustavo/dev-ncar-candidate-2018-09-26


which I can then checkout or merge or whatever.


Kate

**Comment by alperaltuntas** [comment]
To add to Kate's comment, the following is usually what I do to check out a specific PR:

git fetch [remote-name] pull/855/head:pr855
git checkout pr855

where, 855 is the PR number, and so should change for each PR. This will create a local branch called pr855, so you may want to clean those branches from time to time.

**Comment by kshedstrom** [comment]
As for the PR, still no problem from my end.

**Pull request merged**